src/foo/imp.rs

```diff
@@ -105,71 +101,70 @@ impl ObjectImpl for Foo {
         PROPERTIES.as_ref()
     }
 
-    fn set_property(
-        &self,
-        obj: &Self::Type,
-        _id: usize,
-        value: &glib::Value,
-        pspec: &glib::ParamSpec,
-    ) {
+    fn set_property(&self, _id: usize, value: &glib::Value, pspec: &glib::ParamSpec) {
         match pspec.name() {
             "name" => {
                 let name = value.get().unwrap();
-                self.set_name(obj, name);
+                self.set_name(name);
             }
             _ => unimplemented!(),
         }
     }
 
-    fn property(&self, obj: &Self::Type, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
+    fn property(&self, _id: usize, pspec: &glib::ParamSpec) -> glib::Value {
         match pspec.name() {
-            "name" => self.name(obj).to_value(),
+            "name" => self.name().to_value(),
             _ => unimplemented!(),
         }
     }
 }
 
 impl NameableImpl for Foo {
-    fn name(&self, nameable: &Self::Type) -> Option<String> {
-        self.name(nameable.dynamic_cast_ref().unwrap())
+    fn name(&self) -> Option<String> {
+        self.name()
     }
 }
 
 impl Foo {
     //
     // Safe implementations. These take the wrapper type, and not &Self, as first argument
     //
-    fn increment(&self, this: &super::Foo, inc: i32) -> i32 {
+    fn increment(&self, inc: i32) -> i32 {
         let mut val = self.counter.borrow_mut();
 
         *val += inc;
 
-        this.emit_by_name::<()>("incremented", &[&*val, &inc]);
+        self.obj().emit_by_name::<()>("incremented", &[&*val, &inc]);
 
         *val
     }
 
-    fn incremented(&self, _this: &super::Foo, _val: i32, _inc: i32) {
+    fn incremented(&self, _val: i32, _inc: i32) {
         // Could do something here. Default/class handler of the "incremented"
         // signal that could be overriden by subclasses
     }
 
-    fn counter(&self, _this: &super::Foo) -> i32 {
+    fn counter(&self) -> i32 {
         *self.counter.borrow()
     }
 
-    fn name(&self, _this: &super::Foo) -> Option<String> {
+    fn name(&self) -> Option<String> {
         self.name.borrow().clone()
     }
 
-    fn set_name(&self, _this: &super::Foo, name: Option<String>) {
+    fn set_name(&self, name: Option<String>) {
         *self.name.borrow_mut() = name;
     }
+
+    async fn check_async(&self) -> Result<(), &'static str> {
```

Comment:
```suggestion
    async fn check_async(&self) -> Result<(), glib::Error> {
```

would be more consistent or not?

Reply:
yeah, that makes sense.